Core/Src/main.c

```diff
@@ -582,12 +582,40 @@ void HAL_DAC_ConvCpltCallbackCh1(DAC_HandleTypeDef *hdac) {
 	HAL_DAC_Stop_DMA(&hdac1, DAC_CHANNEL_1);
 	if (toneIndex >= toneSequenceSize) {
 		playingSequence = 0;
+		toneIndex = 0;
+		HAL_DFSDM_FilterRegularStart_DMA(&hdfsdm1_filter0, currentTone, TONE_LEN);
 		return;
 	}
 	toneIndex++;
 	loadToneFromFlash(toneSequence[toneIndex]);
 	HAL_DAC_Start_DMA(&hdac1, DAC_CHANNEL_1, currentTone, TONE_LEN, DAC_ALIGN_8B_R);
 }
+
+// callback for timer
+void HAL_DFSDM_FilterRegConvCpltCallback (DFSDM_Filter_HandleTypeDef * hdfsdm_filter)
+{
+	// check to make sure correct callback
+	if(hdfsdm_filter == &hdfsdm1_filter0)
+	{
+		HAL_DFSDM_FilterRegularStop_DMA(hdfsdm_filter);
+
+		// scale 24 bits to 8 bit channel
+		for(size_t i = 0; i < TONE_LEN; i++)
+		{
+			currentTone[i] = (currentTone[i] >> 8)/65536 + 128;
+		}
+
+		//write microphone to flash
+		if (BSP_QSPI_Write(currentTone, sizeof(currentTone)*NUM_TONES + toneIndex * TONE_LEN, sizeof(currentTone)) != QSPI_OK) {
+			Error_Handler();
+		}
+
+		if(++toneIndex < toneSequenceSize)
+		{
+			HAL_DFSDM_FilterRegularStart_DMA(&hdfsdm1_filter0, currentTone, TONE_LEN);

```

Comment:
this is not particularly important for THIS PR, but I was thinking looking at this code that there should be a flag we set sort of like the `playingSequence` flag (perhaps `recordingSequence`) that lets other parts of the code see when the users input has been fully recorded, or is in the process of being recorded. Now that i think about it, we could use the main loop to run asyncronous processing of the recorded tones as soon as they hit flash, which would involve complexity outside of the scope of the PR. just sharing my thoughts as i read the code lol no changes needed

Reply:
Interesting, the problem with the main loop doing asynchronous processing is we would always have to poll a flag for when to process stuff. I think it would be better to have this in some sort of callback or ISR that is done immediately after getting all input